File: c/luhn/luhn.c
#include "luhn.h"
bool luhn(const char *num){

	//final sum, length of array, cound of digits
	int sum = 0, l = 0, c = 0;
	//set to determine if digit should be doubled
	bool dbl_flg = false;
	
	//Gets the length of the string without the string library
	while(num[l] != '\0'){
		l++;
	}
		
	//If the length of the array is less than 2, don't event test
	if(l < 2) return false;
	
	for(int i = l - 1; i >= 0; i--)
	{
		//Itterate loop if character is white space
		if(num[i] == ' ') continue;
		//Only consider characters 0-9
		if(num[i] <= '9' && num[i] >= '0'){
			//Add 1 to count of digits
			c++;
			if(!dbl_flg){
				sum += num[i] - '0';
				dbl_flg = true;
			} else if(dbl_flg){
				int s1 = (num[i] - '0') *2;
				dbl_flg = false;
				if(s1 > 9){
					sum += (s1-9);
				}else{
					sum+= s1;
				}
			}
		}else{
			return false;
		}
	}
	return (sum % 10 == 0) && c >= 2;
}

File: go/luhn/luhn.go
package luhn

import (
	"strings"
)

func Valid(str string) bool {
	//remove white space
	str = strings.Join(strings.Fields(str), "")
	//initialize sum
	sum := 0
	//used to determine if string is na odd length. This determines whether even or odd numbers are doubled when itterating from left to right
	odd := len(str) % 2

	//Must have at least 2 digits
	if len(str) < 2 {
		return false
	}
	// for each number in the string...
	for i, r := range str {
		//cast the rune as an interger but subtract the rune value of 0 so it reflects the proper digit
		digit := int(r - '0')
		//if digit is a value that does not correspond to the ascii value of a digit, fail
		if digit > 9 {
			return false
		}
		//if number of digits is odd and if we're on an odd digit... (counting starts at 0)
		if i%2 == odd {
			//multiply number by 2
			digit *= 2
			//if the number is greater than 9...
			if digit > 9 {
				//subtract 9 from the number
				digit -= 9
			}
		}
		//add the digit to the sum
		sum += digit
	}
	//return true if the sum is evenly divisible by 10
	return sum%10 == 0
}
